lib/api/src/module.rs

```diff
@@ -411,14 +414,40 @@ impl Module {
         &self.artifact.module_ref()
     }
 
-    /// Gets the [`Artifact`] used internally by the Module.
+    /// Query the artifact to fetch information from it without
+    /// extracting the artifact from the module.
+    ///
+    /// The artifact defined in the `Module` is abstracted behind `dyn
+    /// Artifact`. When querying, downcasting to a concrete artifact
+    /// type is done by inferring the type from the closure used to
+    /// query. Something like this:
+    ///
+    /// ```rust,no_run
+    /// use wasmer_engine_object_file::ObjectFileArtifact;
+    ///
+    /// let symbol_registry = unsafe {
+    ///     module.query_artifact(|artifact: &ObjectFileArtifact| {
+    ///         Some(Arc::clone(artifact.symbol_registry()))
+    ///     })
+    /// };
+    /// ```
+    ///
+    /// If downcasting the artifact to `A` fails, the method will
+    /// return `None`.
     ///
     /// This API is hidden because it's not necessarily stable;
     /// this functionality is required for some core functionality though, like
-    /// the object file engine.
+    /// the [`wasmer-engine-object-file`].
     #[doc(hidden)]
-    pub fn artifact(&self) -> &Arc<dyn Artifact> {
-        &self.artifact
+    pub unsafe fn query_artifact<A, Q, R>(&self, query: Q) -> Option<Arc<R>>
+    where
+        A: Artifact + 'static,
+        Q: FnOnce(&A) -> Option<Arc<R>>,
```

Comment:
Returning something always wrapped in an `Arc` seems perhaps overly restrictive... I think we could probably just make `R` also `Cloned`, or does that not work?

Reply:
I wanted to go straight to the point. Setting it to `R` means we can't return a reference, that's annoying. We should not want to implement `Clone` for `SymbolRegistry`s for instance. Thoughts?

Reply:
Well, my idea is that `Arc` is clone, so we can still always return `Arc` if we want... maybe that makes the function harder to use if users have to specify what they mean directly as a type argument (though perhaps making this function harder to use is a feature 😉 )

Reply:
I think we can just remove this one.